rlclientlib/logger/event_logger.h

```diff
@@ -124,13 +128,34 @@ class multi_slot_logger : public event_logger<multi_slot_decision_event> {
     int report_action_taken(const char* event_id, api_status* status);
   };
 
-  class generic_event_logger : public event_logger<generic_event> {
+  class generic_event_logger : public event_logger<std::function<int(generic_event&, api_status*)>> {
   public:
-    generic_event_logger(i_time_provider* time_provider, i_async_batcher<generic_event>* batcher, const char* app_id)
+    generic_event_logger(i_time_provider* time_provider, i_async_batcher<std::function<int(generic_event&, api_status*)>>* batcher, const char* app_id)
       : event_logger(time_provider, batcher, app_id)
     {}
 
-    int log(const char* event_id, generic_event::payload_buffer_t&& payload, generic_event::payload_type_t type, event_content_type content_type, api_status* status);
-    int log(const char* event_id, generic_event::payload_buffer_t&& payload, generic_event::payload_type_t type, event_content_type content_type, generic_event::object_list_t&& objects, api_status* status);
+    template<typename TSerializer, typename... Args>
+    int log(const char* event_id, const char* context, generic_event::payload_type_t type, i_logger_extensions* ext, TSerializer& serializer, api_status* status, Args... args) {
+      const auto now = _time_provider != nullptr ? _time_provider->gmt_now() : timestamp();
+      // using std::bind for the in_evt to get in-place construction.
+      // We can replace this with a pure lambda in C++14
+      using namespace std::placeholders;
+      std::function<int(generic_event&, api_status*)> evt_fn =
+        std::bind(
+          [type, ext, serializer, args...](generic_event& out_evt, api_status* status, generic_event in_evt)->int
+          {
+            RETURN_IF_FAIL(in_evt.transform(ext, serializer, status, args...));
+            out_evt = std::move(in_evt);
+            return error_code::success;
+          },
+          _1,
+          _2,
+          generic_event(event_id, now, type, context, _app_id)
+        );
+      return append(std::move(evt_fn), event_id, 1, status);
```

Comment:
will all these pointers (`api_status`, `ext`) be now used in an asynchronous manner, withought locking, by both threads? I am wondering this will cause a problem? i.e. thread 1 sets status to fail, thread 2 sets status to success, thread 1 returns now incorrect status

Reply:
This workflow already exists in the code. I'm not sure if its safe, but it would need to be a larger refactor to fix any issues with it